File: ios/DrawingEngine/Sources/DrawingEngine/Model/ShapeTarget.swift
import Foundation

/// The kind of mark a level asks the child to make. Drives which metrics matter
/// and how stars are weighted.
public enum ShapeKind: String, Sendable, Codable, CaseIterable {
    case line
    case arc
    case circle
    case rectangle
    case triangle
    case spiral
    case zigzag
    case wave
}

/// The ideal mark the child is aiming at. Every target can produce its ideal
/// path as a polyline, which is what scoring compares the stroke against.
public enum ShapeTarget: Sendable, Equatable {
    case line(from: Point2D, to: Point2D)
    case circle(center: Point2D, radius: Double)
    /// Arbitrary path; `closed` joins the last point back to the first.
    case polyline(points: [Point2D], closed: Bool)

    public var isClosed: Bool {
        switch self {
        case .line: return false
        case .circle: return true
        case .polyline(_, let closed): return closed
        }
    }

    /// The ideal path as evenly described points (closed shapes repeat the first point at the end).
    public func idealPath(samples: Int = 128) -> [Point2D] {
        switch self {
        case .line(let a, let b):
            let n = max(2, samples)
            return (0..<n).map { i in
                let t = Double(i) / Double(n - 1)
                return Point2D(a.x + (b.x - a.x) * t, a.y + (b.y - a.y) * t)
            }
        case .circle(let center, let radius):
            let n = max(8, samples)
            return (0...n).map { i in
                let angle = 2 * Double.pi * Double(i) / Double(n)
                return Point2D(center.x + radius * cos(angle),
                               center.y + radius * sin(angle))
            }
        case .polyline(let points, let closed):
            if closed, let first = points.first { return points + [first] }
            return points
        }
    }

    public var boundingBox: BoundingBox? { BoundingBox.of(idealPath()) }

    /// A stable size used to normalize positional error into a 0–1 score.
    public var characteristicLength: Double {
        max(boundingBox?.diagonal ?? 1, 1e-6)
    }

    // MARK: - Builders for parametric shapes (used by content definitions)

    public static func arc(center: Point2D, radius: Double,
                           startAngle: Double, endAngle: Double, samples: Int = 64) -> ShapeTarget {
        let n = max(2, samples)
        let pts = (0..<n).map { i -> Point2D in
            let t = Double(i) / Double(n - 1)
            let angle = startAngle + (endAngle - startAngle) * t
            return Point2D(center.x + radius * cos(angle), center.y + radius * sin(angle))
        }
        return .polyline(points: pts, closed: false)
    }

    public static func rectangle(_ box: BoundingBox) -> ShapeTarget {
        .polyline(points: [
            Point2D(box.minX, box.minY),
            Point2D(box.maxX, box.minY),
            Point2D(box.maxX, box.maxY),
            Point2D(box.minX, box.maxY)
        ], closed: true)
    }

    public static func triangle(_ a: Point2D, _ b: Point2D, _ c: Point2D) -> ShapeTarget {
        .polyline(points: [a, b, c], closed: true)
    }

    public static func spiral(center: Point2D, startRadius: Double, endRadius: Double,
                              turns: Double, samples: Int = 160) -> ShapeTarget {
        let n = max(2, samples)
        let pts = (0..<n).map { i -> Point2D in
            let t = Double(i) / Double(n - 1)
            let angle = 2 * Double.pi * turns * t
            let radius = startRadius + (endRadius - startRadius) * t
            return Point2D(center.x + radius * cos(angle), center.y + radius * sin(angle))
        }
        return .polyline(points: pts, closed: false)
    }
}

File: web/src/lib/engine/shapeTarget.ts
import type { BoundingBox, Point2D } from './geometry'
import { bboxDiagonal, boundingBox } from './geometry'

export type ShapeKind =
  | 'line' | 'arc' | 'circle' | 'rectangle' | 'triangle' | 'spiral' | 'zigzag' | 'wave'

export type ShapeTarget =
  | { readonly kind: 'line'; readonly from: Point2D; readonly to: Point2D }
  | { readonly kind: 'circle'; readonly center: Point2D; readonly radius: number }
  | { readonly kind: 'polyline'; readonly points: readonly Point2D[]; readonly closed: boolean }

export function isClosed(t: ShapeTarget): boolean {
  switch (t.kind) {
    case 'line': return false
    case 'circle': return true
    case 'polyline': return t.closed
  }
}

/** The ideal mark, sampled as a polyline. Closed shapes repeat the first point at the end. */
export function idealPath(t: ShapeTarget, samples: number = 128): Point2D[] {
  switch (t.kind) {
    case 'line': {
      const n = Math.max(2, samples)
      const out: Point2D[] = []
      for (let i = 0; i < n; i++) {
        const u = i / (n - 1)
        out.push({
          x: t.from.x + (t.to.x - t.from.x) * u,
          y: t.from.y + (t.to.y - t.from.y) * u
        })
      }
      return out
    }
    case 'circle': {
      const n = Math.max(8, samples)
      const out: Point2D[] = []
      for (let i = 0; i <= n; i++) {
        const a = (2 * Math.PI * i) / n
        out.push({
          x: t.center.x + t.radius * Math.cos(a),
          y: t.center.y + t.radius * Math.sin(a)
        })
      }
      return out
    }
    case 'polyline': {
      return t.closed && t.points.length > 0
        ? [...t.points, t.points[0]]
        : [...t.points]
    }
  }
}

export function targetBoundingBox(t: ShapeTarget): BoundingBox | null {
  return boundingBox(idealPath(t))
}

/** A stable size used to normalise positional error to 0..1 scores. */
export function characteristicLength(t: ShapeTarget): number {
  const bb = targetBoundingBox(t)
  return bb ? Math.max(bboxDiagonal(bb), 1e-6) : 1e-6
}

// Builders for parametric shapes (used by content definitions).

export function arc(
  center: Point2D, radius: number, startAngle: number, endAngle: number, samples: number = 64
): ShapeTarget {
  const n = Math.max(2, samples)
  const points: Point2D[] = []
  for (let i = 0; i < n; i++) {
    const u = i / (n - 1)
    const a = startAngle + (endAngle - startAngle) * u
    points.push({ x: center.x + radius * Math.cos(a), y: center.y + radius * Math.sin(a) })
  }
  return { kind: 'polyline', points, closed: false }
}

export function rectangle(box: BoundingBox): ShapeTarget {
  return {
    kind: 'polyline',
    closed: true,
    points: [
      { x: box.minX, y: box.minY },
      { x: box.maxX, y: box.minY },
      { x: box.maxX, y: box.maxY },
      { x: box.minX, y: box.maxY }
    ]
  }
}

export function triangle(a: Point2D, b: Point2D, c: Point2D): ShapeTarget {
  return { kind: 'polyline', points: [a, b, c], closed: true }
}

export function spiral(
  center: Point2D, startRadius: number, endRadius: number, turns: number, samples: number = 160
): ShapeTarget {
  const n = Math.max(2, samples)
  const points: Point2D[] = []
  for (let i = 0; i < n; i++) {
    const u = i / (n - 1)
    const a = 2 * Math.PI * turns * u
    const r = startRadius + (endRadius - startRadius) * u
    points.push({ x: center.x + r * Math.cos(a), y: center.y + r * Math.sin(a) })
  }
  return { kind: 'polyline', points, closed: false }
}
